include/ensmallen_bits/callbacks/report.hpp

```diff
@@ -0,0 +1,558 @@
+/**
+ * @file report.hpp
+ * @author Marcus Edel
+ *
+ * Implementation of a simple report callback function.
+ *
+ * ensmallen is free software; you may redistribute it and/or modify it under
+ * the terms of the 3-clause BSD license.  You should have received a copy of
+ * the 3-clause BSD license along with ensmallen.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef ENSMALLEN_CALLBACKS_REPORT_HPP
+#define ENSMALLEN_CALLBACKS_REPORT_HPP
+
+#include <ensmallen_bits/function.hpp>
+#include <iomanip>
+
+namespace ens {
+
+/**
+ * A simple optimization report.
+ */
+class Report
+{
+ public:
+  /**
+   * Set up the report callback class with the given output stream.
+   *
+   * @param iterationsPercentageIn The number of iterations to report in
+   *     percent, between [0, 1]).
+   * @param outputIn Ostream which receives output from this object.
+   * @param outputMatrixSizeIn The number of values to output for the function
+   *     coordinates.
+   */
+  Report(const double iterationsPercentageIn = 0.1,
+         std::ostream& outputIn = arma::get_cout_stream(),
+         const size_t outputMatrixSizeIn = 4) :
+      iterationsPercentage(iterationsPercentageIn),
+      output(outputIn),
+      outputMatrixSize(outputMatrixSizeIn),
+      objective(0),
+      gradientNorm(0),
+      hasGradient(false),
+      hasEndEpoch(false),
+      gradientCalls(0),
+      evaluateCalls(0)
+  { /* Nothing to do here. */ }
+
+  /**
+   * Callback function called at the begin of the optimization process.
+   *
+   * @param optimizer The optimizer used to update the function.
+   * @param function Function to optimize.
+   * @param coordinates Starting point.
+   */
+  template<typename OptimizerType, typename FunctionType, typename MatType>
+  void BeginOptimization(OptimizerType& /* optimizer */,
+                         FunctionType& /* function */,
+                         MatType& coordinates)
+  {
+    initialCoordinates = coordinates;
+    optimizationTimer.tic();
+  }
+
+  /**
+   * Callback function called at the begin of the optimization process.
+   *
+   * @param optimizer The optimizer used to update the function.
+   * @param function Function to optimize.
+   * @param coordinates Starting point.
+   */
+  template<typename OptimizerType, typename FunctionType, typename MatType>
+  void EndOptimization(OptimizerType& optimizer,
+                       FunctionType& function,
+                       MatType& coordinates)
+  {
+    output << "Report" << std::endl;
```

Comment:
```suggestion
    output << "Optimization Report" << std::endl;
```
I dunno, up to you if you want to change this.  Maybe it's a little more clear? :)